sql/analyzer/costed_index_scan_test.go

```diff
@@ -0,0 +1,201 @@
+// Copyright 2023 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package analyzer
+
+import (
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/expression"
+)
+
+func TestExprFlatten(t *testing.T) {
+	tests := []struct {
+		name     string
+		in       sql.Expression
+		exp      string
+		leftover string
+	}{
+		{
+			name: "ands",
+			in: and(
+				and(
+					eq(gf(0, "xy", "x"), lit(1)),
+					eq(gf(1, "xy", "y"), lit(2)),
+				),
+				gt(gf(1, "xy", "y"), lit(0)),
+			),
+			exp: `
+(1: and
+  (3: xy.x = 1)
+  (4: xy.y = 2)
+  (5: xy.y > 0))`,
+		},
+		{
+			name: "ands with or",
+			in: and(
+				and(
+					eq(gf(0, "xy", "x"), lit(1)),
+					eq(gf(1, "xy", "y"), lit(2)),
+				),
+				or(
+					gt(gf(1, "xy", "y"), lit(0)),
+					lt(gf(1, "xy", "y"), lit(7)),
+				),
+			),
+			exp: `
+(1: and
+  (3: xy.x = 1)
+  (4: xy.y = 2)
+  (5: or
+    (7: and
+      (6: xy.y > 0))
+    (8: and
+      (7: xy.y < 7))))`,
+		},
+		{
+			name: "ors with and",
+			in: or(
+				or(
+					eq(gf(0, "xy", "x"), lit(1)),
+					eq(gf(1, "xy", "y"), lit(2)),
+				),
+				and(
+					gt(gf(0, "xy", "y"), lit(0)),
+					lt(gf(0, "xy", "y"), lit(7)),
+				),
+			),
+			exp: `
+(1: or
+  (4: and
+    (3: xy.x = 1))
+  (5: and
+    (4: xy.y = 2))
+  (5: and
+    (6: xy.y > 0)
+    (7: xy.y < 7)))`,
```

Comment:
Why do we create the parent `and` expressions on line 85 and 87 over the `or`s? In particular, it seemed odd that `5: and` contained expression child number `4` when it doesn't seem to in the original expression. 

I was expecting a flattened form more like:
```
(1: or
  (3: xy.x = 1))
  (4: xy.y = 2))
  (5: and
    (6: xy.y > 0)
    (7: xy.y < 7)))`,
```

Reply:
I think the number overlaps were getting in the way of filter pruning, index_query_plans.go is now -11k